sql/sem: extract eval from tree

ajwerner opened the pull request:
This massive commit is a pure refactor: no logic should be changed by it.

### Background

Cockroach, from its earliest days, has overloaded the tree package with
functionality additionally to type check and evaluate expressions represented
in the AST. The big downside here is that it becomes hard to isolate the tree
package from all of the dependencies of evaluation. What we want is to be able
to parse cockroach SQL without needing to depend on much at all. This was an
increasingly large hurdle.

Getting `tree` out of the business of type checking and some of the comparison
business is a big task not handled here. It might be nice to do one day. In
some ways it's harder. We get a lot of efficiency benefit by being able to
store this type information and data awareness in the ast node objects
directly. If we were to want to transform during type-checking, we may need
to incur overhead due to allocations. That's all left for future work.

### Nuts and bolts

This PR creates a number of new packages:

 * sql/sem/volatility
    * Tiny package, defines volatilities.
 * sql/sem/internal/cast
    * Defines rules for casting which are used by both type checking and eval.
 * sql/sem/eval
    * Implements expression and operation evaluation.
    * Moves the `tree.EvalContext` => `eval.Context`.
 * sql/sem/normalize
    * Takes the process of normalization and its use of constant evaluation
      into a new package.
 * sql/sem/tree/evalgen
    * Code generator for the interfaces discussed below.

In terms of the mechanics of this change, some new gigantic interfaces have
been introduced: ExprEvaluator and OpEvaluator. These interfaces and generated
instance methods are a classic example of the GoF's
https://en.wikipedia.org/wiki/Visitor_pattern.

In effect, the pattern trades a double indirection for a sealed mechanism to
add new functionality in the eval package to structs defined in tree. Time
will tell whether this is an efficiency problem. If it is, it's not too hard
to convert the dispatch to a big switch.

The code generator just uses go ast introspection to drive the generator. Some
of the ways in interprets the set of types to use is perhaps ad-hoc, but it's
concrete and precise enough.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80504)
<!-- Reviewable:end -->

jordanlewis commented:
@ajwerner is there any way you could break this up into a PR defining the new package, then a PR that just s/EvalContext/eval.Context/ or something? Maybe that wouldn't work on its own. Just wanting to ask if there's a way you can split this up a little bit to help with review and later inspection (without adding, hopefully, too much extra work on your end).

jordanlewis commented:
BTW, I should have said up front, this is incredible work and I'm extremely excited about it - thank you very much for taking the initiative!

ajwerner commented:
When I was doing the programming, the unravelling did not happen in a tidy way; one commit was not worse than the git history I had accumulated. I was eager enough to post it. There's probably a few low-effort packages to pull out up front like the `volatility` and `cast` packages, but once the adoption gets going, it'll be some work to re-arrange it. On some level, I'd rather @otan or, if we anoint any other reviewers to engage on how much splitting commits would actually help. I'm going to spend some time pairing with him tomorrow in his morning, and will take next steps from there. 

The truth is that there's two-ish sections (the ops and exprs) of manual moving of code from one place to another that had the opportunity for bugs to be introduced (indeed, in one place I swapped a `right` and `left` and fortunately a bunch of tests broke). Also fortunately, it's just a sort of straight-up code movement. I think we pay attention to those things, much of the rest of this will just scroll by without anything to raise an eyebrow over. FWIW, this is much less complex than was https://github.com/cockroachdb/cockroach/pull/52519. It was much more just rote restructuring and figuring out some good little sed snippets than that one which really needed to be carefully decomposed. 

One of the areas I feel most likely to have left room for improvement on the table is an area not really affected by this code: there's tests in the `tree_test` package that probably should be disseminated but I left in tree for now because I was done hunting for code to move.

ajwerner committed:
sql/sem/tree/evalgen: generate interfaces/methods to eval exprs and ops

This commit does two things:

1) It defines all of the UnaryEvalOp and BinaryEvalOp implementations in their
   respective files. These files both define the operations and are the input
   to the code generator which parses them to produce methods to eval each op
   and the interface of the OpEvaluator to evaluate them.
2) It adds the `evalgen` tool which parses the tree library files and generates
   the above-mentioned operation interfaces and a parallel set of methods and
   an interface for expressions.

Release note: None

ajwerner committed:
sql/sem/eval: extract a package from tree

Release note: None

ajwerner commented:
Okay, I broke it out into 7 commits. The second-to-last is the big one.

otan commented:
i've looked at the first four commits and i'd say you're good to merge those, if you wish

RichardJCai commented on pkg/sql/colexec/colexecprojconst/BUILD.bazel:
```diff
@@ -26,7 +26,8 @@ go_library(
         "//pkg/sql/execinfra/execreleasable",  # keep
-        "//pkg/sql/sem/tree",
+        "//pkg/sql/sem/eval",
+        "//pkg/sql/sem/tree",  # keep
```
Random question, what does the `# keep` do here?

RichardJCai commented on pkg/sql/sem/eval/overload.go:
```diff
@@ -0,0 +1,96 @@
+type FnOverload = func(*Context, tree.Datums) (tree.Datum, error)
+
+// FnWithExprsOverload is the concrete type for the tree.Overload.FnWithExprs
+//// field.
```
nit extra slashes here.

ajwerner replied:
Done.

RichardJCai reviewed (approved):
Bravo, this is amazing work.
I looked at the last commit which looks generally good to me.

I think there's also some rebase skew, idk if you fixed it during my review

```
pkg/sql/crdb_internal.go:6022:16: undefined: tree.TimestampToInexactDTimestamp
pkg/sql/crdb_internal.go:6034:16: undefined: tree.TimestampToInexactDTimestamp
```

Also gotta say I skipped looking at all the `_test.go` files

ajwerner commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -1382,7 +1383,7 @@ func (dsp *DistSQLPlanner) maybeParallelizeLocalScans(
 	// then we will split all spans according to the leaseholder boundaries and
 	// will create a separate TableReader for each node.
-	sd := planCtx.ExtendedEvalCtx.EvalContext.SessionData()
+	sd := planCtx.ExtendedEvalCtx.Context.SessionData()
```
Done.

ajwerner commented on pkg/sql/sem/builtins/pg_builtins.go:
```diff
@@ -1721,39 +1722,39 @@ SELECT description
+				return eval.ObjectNotFound, nil
 			}
 			// All users have CREATE privileges in all tablespaces.
 			_ = privs
```
I think it's just there as a conduit for commentary and to suppress the unused variable compiler error. I'm not sure why it was done that way.

ajwerner commented on pkg/sql/sem/eval/context.go:
```diff
@@ -0,0 +1,571 @@
+	return out
+}
+
+// MakeTestingEvalContext returns an EvalContext that includes a MemoryMonitor.
```
Let's do it in a follow-up

RichardJCai replied:
I'm happy with doing everything I mentioned here in a follow up

ajwerner commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -724,18 +739,18 @@ func (d *DInt) Compare(ctx *EvalContext, other Datum) int {
 	thisInt := *d
 	var v DInt
-	switch t := UnwrapDatum(ctx, other).(type) {
+	switch t := du.UnwrapDatum(other).(type) {
```
There was an intermediate interface named `DatumUnwrapper` which preceded the `CompareContext` in the course of the refactor. Moved them all to `ctx`

ajwerner commented on pkg/sql/sem/eval/expr.go:
```diff
@@ -0,0 +1,609 @@
+
+func (e *evaluator) ctx() *Context { return (*Context)(e) }
+
+func (e *evaluator) EvalAllColumnsSelector(selector *tree.AllColumnsSelector) (tree.Datum, error) {
```
I suppose we could move the method implementation. This mirrors the logic in the pre-existing eval method. I'm inclined to leave it for now.

otan reviewed (commented):
I've had a 1:1 discussing the changes, at a high level they LGTM. I took a look at only a few files (tree and eval ones) and they all look good to me.

I have minor comments which can be addressed later. If this is green, I'll rebase and merge (there's already a conflict) in my arvo.

ajwerner commented on pkg/sql/sem/tree/hlc.go:
```diff
@@ -0,0 +1,79 @@
+// Copyright 2018 The Cockroach Authors.
```
Technically, I moved stuff out of the old file (into asof/asof.go) and then I renamed the file because all it has was HLC stuff in it.

ajwerner replied:
This is like theseus's ship, but before the last planks had even been replaced.

ajwerner commented on pkg/sql/sem/tree/eval_binary_ops.go:
```diff
@@ -0,0 +1,371 @@
+type ContainsArrayOp struct{}
+
+// ContainsJsonbOp is a BinaryEvalOp.
+type ContainsJsonbOp struct{}
```
Yeah, however the type symbol is `types.Jsonb`, so I went with that here.

otan replied:
there's some ops capitalised as JSONB though, IIRC. oh well, no big deal for now

otan commented:
bors r+ p=9999 single on

let's start mergin'

otan commented:
Flake

bors r+ p=9999 single on
